### PR TITLE
RGW - Zipper - Allow per-store admin APIs

### PR DESCRIPTION
**Pull request by dang**
Some of the existing admin APIs are rados-specific, and other stores will want to add specific APIs.  Allow this by having a function to add store-specific APIs.

Signed-off-by: Daniel Gryniewicz <dang@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by dang** [comment]
https://pulpito.ceph.com/dang-2022-09-06_14:33:01-rgw-wip-dang-zipper-admin-api-distro-default-smithi/

Good run.

File: src/rgw/rgw_sal_rados.cc [hunk 1]
```diff
@@ -1420,6 +1427,18 @@ void RadosStore::finalize(void)
     rados->finalize();
 }
 
+void RadosStore::register_admin_apis(RGWRESTMgr_Admin* mgr)
+{
+  mgr->register_resource("bucket", new RGWRESTMgr_Bucket);
+  /*Registering resource for /admin/metadata */
+  mgr->register_resource("metadata", new RGWRESTMgr_Metadata);
+  mgr->register_resource("log", new RGWRESTMgr_Log);
```
Comment (cbodley) [inline src/rgw/rgw_sal_rados.cc:1435]: these will be required for multisite support in all stores

File: src/rgw/rgw_sal_rados.cc [hunk 2]
```diff
@@ -1420,6 +1427,18 @@ void RadosStore::finalize(void)
     rados->finalize();
 }
 
+void RadosStore::register_admin_apis(RGWRESTMgr_Admin* mgr)
+{
+  mgr->register_resource("bucket", new RGWRESTMgr_Bucket);
```
Comment (cbodley) [inline src/rgw/rgw_sal_rados.cc:1432]: what bucket apis are rados-specific? maybe the link/unlink stuff?

**Comment by dang** [comment]
jenkins test make check

File: src/rgw/rgw_sal_rados.cc (hunk 1 above)
Comment (dang) [reply src/rgw/rgw_sal_rados.cc:1435]: We can move things around once there's multiple implementations.  At the moment, building modules is my goal.

File: src/rgw/rgw_sal_rados.cc (hunk 2 above)
Comment (dang) [reply src/rgw/rgw_sal_rados.cc:1432]: Currently link, unlink, and get stale instances.  At some point, we can fix/break these up.

File: src/rgw/rgw_sal.h [hunk 3]
```diff
@@ -451,6 +452,8 @@ class Store {
     virtual const std::string& get_luarocks_path() const = 0;
     /** Set the location of where lua packages are installed */
     virtual void set_luarocks_path(const std::string& path) = 0;
+    /** Register admin APIs unique to this store */
+    virtual void register_admin_apis(RGWRESTMgr_Admin* mgr) = 0;
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:456]: nit: this can take the base `RGWRESTMgr*`

**Review by cbodley — APPROVED** [review]

**Review by soumyakoduri — APPROVED** [review]

**Comment by dang** [comment]
jenkins test make check

**Comment by dang** [comment]
jenkins test make check

**Pull request merged**